spot-contracts/contracts/vaults/RolloverVault.sol

```diff
@@ -334,6 +359,116 @@ contract RolloverVault is
         return redemptions;
     }
 
+    /// @notice Swaps the bond tranche slices from the user into the underlying collateral for a fee,
+    ///         given that the vault has the remaining slices for immature redemption.
+    /// @dev When a user has some tranche slices from a given bond, and the vault has the remaining slices,
+    ///      this method allows the user and the vault to "meld" their assets together to redeem the underlying collateral
+    ///      from the bond. The user's share of the underlying is returned back and the vault charges
+    ///      the user a fee for providing liquidity.
+    /// @param bond The bond whose tranches are to be swapped.
+    /// @param trancheAmtsIn A list of amounts of each bond tranche the user is to deposit.
+    ///                      When the user does not have a particular slice, its amount in the list is to be set to zero.
+    /// @return The amount of underlying tokens returned.
+    function meld(IBondController bond, uint256[] memory trancheAmtsIn)
+        external
+        nonReentrant
+        whenNotPaused
+        returns (uint256)
+    {
+        // get bond tranches
+        BondTranches memory bt = bond.getTranches();
+
+        // * We check that the bond's underlying collateral token matches the vault's underlying.
+        // * We validate that the bond has at least 2 tranches to perform the meld operation.
+        // * We validate that the bond hasn't reached maturity as matching only works through immature redemption.
+        if (bond.collateralToken() != address(underlying) || bt.tranches.length <= 1 || bond.isMature()) {
+            revert InvalidBond(bond);
+        }
+
+        // We check if the given bond is NOT malicious by examining if:
+        // * The parent bond of all the children tranches is the given bond.
+        // * At least one of the bond's tranches is already in the vault from previous deployments.
+        bool isValidBond = false;
+        for (uint8 i = 0; i < bt.tranches.length; i++) {
+            if (bt.tranches[i].bond() != address(bond)) {
+                revert InvalidBond(bond);
+            }
+
+            isValidBond = isValidBond || _deployed.contains(address(bt.tranches[i]));
+        }
+        if (!isValidBond) {
+            revert InvalidBond(bond);
+        }
+
+        // track the vault's TVL before the match operation
+        uint256 tvlBefore = getTVL();
+
+        // Fist we check if the vault has all the tranches for proportional redemption and if so recover the underlying.
+        // On successful redemption, The vault's individual tranche balances will decrease, and the underlying balances will increase.
+        // If the vault tranches are not eligible for proportional redemption, its a no-op.
+        // NOTE: We sync balances finally at the end of this function.
+        _execImmatureTrancheRedemption(bond, bt);
```

Comment:
Right

Reply:
Got it. This prioritizes AMPL holdings w/in the vault over income from melding, in some cases where the early-redeem redeems tranches that would have otherwise been melded. I don't think this would be too frequent.

The melder will pay a bit more gas than they would otherwise, to support system upkeep.

I think both of these are ok.